static/js/narrow.js

```diff
@@ -182,18 +184,55 @@ exports.activate = function (raw_operators, opts) {
     // Populate the message list if we can apply our filter locally (i.e.
     // with no backend help) and we have the message we want to select.
     if (narrow_state.get_current_filter().can_apply_locally()) {
+        if (select_strategy.flavor === 'first_unread') {
+            var unread_info = narrow_state.get_first_unread_info();
+
+            if (unread_info.flavor === 'found') {
+                select_strategy = {
+                    flavor: 'exact',
+                    msg_id: unread_info.msg_id,
+                };
+            } else if (unread_info.flavor === 'not_found') {
+                if (message_list.all.fetch_status.has_found_newest()) {
+                    select_strategy = {
+                        flavor: 'last_id',
+                    };
+                    activate_status = 'ready_to_preload';
+                }
+            } else {
+                // flavor is cannot_compute
+                activate_status = 'pending_server';
+            }
+        }
+
         if (select_strategy.flavor === 'exact') {
             var msg_to_select = message_list.all.get(select_strategy.msg_id);
             if (msg_to_select !== undefined) {
-                message_util.add_messages(
-                    message_list.all.all_messages(),
-                    message_list.narrowed,
-                    {delay_render: true});
+                activate_status = 'ready_to_preload';
+            }
+        }
+
+        if (activate_status === 'ready_to_preload') {
```

Comment:
I think `ready_to_preload` might be a lot clearer if we wrote it as `ready_to_use_local_cache`?